vayesta/core/types/wf/fci.py

```diff
@@ -419,17 +443,81 @@ def as_cisdtq(self, c0=None):
+        # Now, unpack... TODO
+
+        # T4: aaaa, aaab, aaba, abaa, abab, bbab, bbba, bbbb
+        # aaaa
+        c4_aaaa_pack = (self.ci[t4addra,0] * t4signa).reshape(ijkl_pairs_a, abcd_pairs_a)
+        # bbbb
+        c4_bbbb_pack = (self.ci[0,t4addrb] * t4signb).reshape(ijkl_pairs_b, abcd_pairs_b)
+        # aaab
+        c4_aaab_pack = np.einsum('i,j,ij->ij', t3signa, t1signb, self.ci[t3addra[:,None], t1addrb])
+        assert(c4_aaab_pack.shape == (ijk_pairs_a * abc_pairs_a, noccb * nvirb))
+        # aabb
+        c4_aabb_pack = np.einsum('i,j,ij->ij', t2signa, t2signb, self.ci[t2addra[:,None], t2addrb])
+        assert(c4_aabb_pack.shape == (ij_pairs_a * ab_pairs_a, ij_pairs_b * ab_pairs_b))
+        # abbb
+        c4_abbb_pack = np.einsum('i,j,ij->ij', t1signa, t3signb, self.ci[t1addra[:,None], t3addrb])
+        assert(c4_aabb_pack.shape == (nocca * nvirb, ijk_pairs_b * abc_pairs_b))
+
+        # Now, unpack... TODO
+
+        # alpha, beta, alpha: Use this longhand code as a sanity check
+        # First find the ijk -> abc excitations, where ijab are alpha, and kc are beta
+        c3_comp = np.zeros((ij_pairs_a * ab_pairs_a, noccb * nvirb))
+        c3_aba = np.zeros((nocca, noccb, nocca, nvira, nvirb, nvira))
+        for d_cnta, doub_inda in enumerate(t2addra):
+            ij_a = int(d_cnta / ab_pairs_a)
+            ab_a = d_cnta % ab_pairs_a
+            i, j = ooidx_a[0][ij_a], ooidx_a[1][ij_a] # j ind < i ind
+            a, b = vvidx_a[0][ab_a], vvidx_a[1][ab_a] # b ind < a ind
+            for s_cntb, sing_indb in enumerate(t1addrb):
+                # First index of c3_comp is a compound index of ijab (alpha, alpha) excitations, 
+                # with the second index being the kc (beta, beta) single excitation.
+                c3_comp[d_cnta, s_cntb] = self.ci[doub_inda, sing_indb] * t2signa[d_cnta] * t1signa[s_cntb]
+
+                k = int(s_cntb / nvirb)
+                c = s_cntb % nvirb
+                # Note, we want aba -> aba spin signature, not aab -> aab, which is what we have.
+                # We can therefore swap (jk) and (bc). This does not cause an overall sign change.
+                # We then also want to fill up the contributions between permutations 
+                # amongst the alpha electrons and alpha holes.
+                # If only one is permuted, then this will indeed cause a sign change.
+                assert(i != j)
+                assert(a != b)
+                c3_aba[i,k,j,a,c,b] = c3_comp[d_cnta, s_cntb]
+                c3_aba[j,k,i,a,c,b] = -c3_comp[d_cnta, s_cntb]
+                c3_aba[i,k,j,b,c,a] = -c3_comp[d_cnta, s_cntb]
+                c3_aba[j,k,i,b,c,a] = c3_comp[d_cnta, s_cntb]
+
+        if len(t2addra) > 0:
+            # This is a better way of doing it, and then expand! Put rest in a test.
+            assert(np.allclose(c3_comp, np.einsum('i,j,ij->ij', \
+                t2signa, t1signb, self.ci[t2addra[:,None], t1addrb])))
+        del c3_comp
+
+
```

Comment:
Could some of this be put into other auxiliary functions? Even if we don't use those auxiliary functions anywhere else,
it would help to keep the WaveFunction classes more readable.

Reply:
These functions have slightly more simple syntax now - I think they're fine but I can move their outside of the class if you still have strong opinions